File: 525-contiguous-array/contiguous-array.java
import java.util.*;

class Solution {
    public int findMaxLength(int[] nums) {

        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, -1);

        int count = 0;
        int max_len = 0;

        for (int i = 0; i < nums.length; i++) {

            if (nums[i] == 0) {
                count--;
            } else {
                count++;
            }

            if (map.containsKey(count)) {
                int idx = map.get(count);
                max_len = Math.max(max_len, i - idx);
            } else {
                map.put(count, i);
            }
        }

        return max_len;
    }
}

File: 525-contiguous-array/contiguous-array.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        mp = {}
        mp[0] = -1   # diff 0 at index -1

        c = 0
        c1 = 0
        maxlen = 0

        for i in range(len(nums)):
            if nums[i] == 0:
                c += 1
            else:
                c1 += 1

            diff = c1 - c

            if diff in mp:
                idx = mp[diff]
                length = i - idx
                maxlen = max(length, maxlen)
            else:
                mp[diff] = i

        return maxlen
